### PR TITLE
CDAP-9114 Add retry on local dataset operations

### PR DESCRIPTION
**Pull request by Denton-L**
JIRA: https://issues.cask.co/browse/CDAP-9114
Bamboo: http://builds.cask.co/browse/CDAP-RUT847

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java [hunk 1]
```diff
@@ -505,13 +507,24 @@ private DatasetProperties addLocalDatasetProperty(DatasetProperties properties)
...
-      datasetFramework.addInstance(instanceSpec.getTypeName(), instanceId,
-                                   addLocalDatasetProperty(instanceSpec.getProperties()));
+
+      Retries.callWithRetries(new Retries.Callable<Void, IOException>() {
+        @Override
+        public Void call() throws IOException {
+          try {
+            datasetFramework.addInstance(instanceSpec.getTypeName(), instanceId,
+                                         addLocalDatasetProperty(instanceSpec.getProperties()));
+          } catch (DatasetManagementException e) {
```
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java]: I think we should only retry on the `RetryableException`. This means no need to convert the `DatasetManagementException` to `RetryableException`.

Comment (sagarkapare) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java]: Something like 
``` java
      try {
        Retries.callWithRetries(new Retries.Callable<Void, Exception>() {
          @Override
          public Void call() throws IOException, DatasetManagementException {
            datasetFramework.addInstance(instanceSpec.getTypeName(), instanceId,
                                         addLocalDatasetProperty(instanceSpec.getProperties()));
            return null;
          }
        }, RetryStrategies.fixDelay(Constants.Retry.RUN_RECORD_UPDATE_RETRY_DELAY_SECS, TimeUnit.SECONDS));
      } catch (IOException | DatasetManagementException e) {
        throw e;
      } catch (Exception e) {
        // should not happen
        throw new RuntimeException(e);
      }
```

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java [hunk 2]
```diff
@@ -505,13 +507,24 @@ private DatasetProperties addLocalDatasetProperty(DatasetProperties properties)
...
+            throw new RetryableException(e);
+          }
+          return null;
+        }
+      }, RetryStrategies.fixDelay(Constants.Retry.RUN_RECORD_UPDATE_RETRY_DELAY_SECS, TimeUnit.SECONDS));
```
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java]: `RUN_RECORD_UPDATE_RETRY_DELAY_SECS` variable is only for the purpose of updating the run records. If we want to use it for local datasets too, then I think we can either add new variable or give different name to it. thoughts?

**Review by sagarkapare — COMMENTED** [review]
few comments.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java (hunk 1 above)
Comment (Denton-L) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java]: Sorry, I didn't realise that `DatasetManagementException` was a child of `RetryableException`.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java (hunk 2 above)
Comment (Denton-L) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java]: I guess we can add a new constant for it.

**Comment by Denton-L** [comment]
Thanks @sagarkapare, I've updated the branch.

File: cdap-common/src/main/java/co/cask/cdap/common/conf/Constants.java [hunk 3]
```diff
@@ -1252,6 +1252,7 @@
     public static final String SERVICE_PREFIX = "service.";
     public static final String FLOW_PREFIX = "flow.";
     public static final int RUN_RECORD_UPDATE_RETRY_DELAY_SECS = 5;
+    public static final int CREATE_LOCAL_DATASET_RETRY_DELAY_SECS = 5;
```
Comment (sagarkapare) [inline cdap-common/src/main/java/co/cask/cdap/common/conf/Constants.java]: I think similar retry logic would be required for deletion of the local datasets too. So we can rename this variable to the `LOCAL_DATASET_OPERATION_RETRY_DELAY_SECS`.

**Comment by sagarkapare** [comment]
LGTM 👍

**Pull request merged**